File: src/main/java/dev/slne/surf/essentials/commands/minecraft/DefaultGamemodeCommand.java
package dev.slne.surf.essentials.commands.minecraft;

import dev.jorel.commandapi.executors.NativeResultingCommandExecutor;
import dev.slne.surf.essentials.commands.EssentialsCommand;
import dev.slne.surf.essentials.utils.EssentialsUtil;
import dev.slne.surf.essentials.utils.color.Colors;
import dev.slne.surf.essentials.utils.permission.Permissions;
import lombok.val;
import net.kyori.adventure.text.Component;
import org.bukkit.GameMode;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.HumanEntity;

public class DefaultGamemodeCommand extends EssentialsCommand {
    public DefaultGamemodeCommand() {
        super("defaultgamemode", "defaultgamemode [<defaultgamemode>]", "Change default gamemode");

        withPermission(Permissions.DEFAULT_GAMEMODE_PERMISSION);

        executesNative((NativeResultingCommandExecutor) (sender, args) -> getGameMode(sender.getCallee()));
        then(gameModeArgument("gamemode")
                .executesNative((NativeResultingCommandExecutor) (sender, args) -> setGameMode(sender.getCallee(), args.getUnchecked("gamemode"))));
    }


    private int setGameMode(CommandSender source, GameMode newGameMode) {
        val server = source.getServer();
        int playersUpdated = 0;

        server.setDefaultGameMode(newGameMode);

        if (EssentialsUtil.properties_forceGamemode()) {
            for (HumanEntity player : server.getOnlinePlayers()) {
                player.setGameMode(newGameMode);
                playersUpdated++;
            }
        }

        EssentialsUtil.sendSuccess(source, Component.text("Der standard Spielmodus wurde auf ", Colors.SUCCESS)
                .append(Component.translatable(newGameMode.translationKey(), Colors.VARIABLE_VALUE))
                .append(Component.text(" gesetzt!", Colors.SUCCESS)));
        return playersUpdated;
    }

    private int getGameMode(CommandSender source) {
        val gameType = source.getServer().getDefaultGameMode();

        EssentialsUtil.sendSuccess(source, Component.text("Der standard Spielmodus ist ", Colors.INFO)
                .append(Component.translatable(gameType.translationKey(), Colors.VARIABLE_VALUE))
                .append(Component.text("!", Colors.INFO)));

        return 1;
    }
}

File: src/main/kotlin/dev/slne/surf/essentials/command/minecraft/DefaultGameModeCommand.kt
package dev.slne.surf.essentials.command.minecraft

import dev.jorel.commandapi.kotlindsl.anyExecutor
import dev.jorel.commandapi.kotlindsl.commandTree
import dev.jorel.commandapi.kotlindsl.getValue
import dev.slne.surf.essentials.command.argument.gameModeArgument
import dev.slne.surf.essentials.util.EssentialsPermissionRegistry
import dev.slne.surf.essentials.util.translatable
import dev.slne.surf.surfapi.core.api.messages.Colors
import dev.slne.surf.surfapi.core.api.messages.adventure.sendText
import org.bukkit.Bukkit
import org.bukkit.GameMode

fun defaultGameModeCommand() = commandTree("defaultgamemode") {
    withPermission(EssentialsPermissionRegistry.GAME_MODE_COMMAND_DEFAULT)
    anyExecutor { executor, _ ->
        executor.sendText {
            appendPrefix()
            info("Der Standard-Spielmodus ist zurzeit auf ")
            translatable(Bukkit.getDefaultGameMode().translationKey()).color(Colors.VARIABLE_VALUE)
            info("gesetzt.")
        }
    }

    gameModeArgument("gameMode") {
        anyExecutor { executor, args ->
            val gameMode: GameMode by args
            val defaultGameMode = Bukkit.getDefaultGameMode()

            if (gameMode == defaultGameMode) {
                executor.sendText {
                    appendPrefix()
                    error("Der Standard-Spielmodus ist bereits auf ")
                    translatable(gameMode.translationKey()).color(Colors.VARIABLE_VALUE)
                    error(" gesetzt.")
                }
                return@anyExecutor
            }

            Bukkit.setDefaultGameMode(gameMode)

            executor.sendText {
                appendPrefix()
                success("Du hast den Standard-Spielmodus auf ")
                translatable(gameMode.translationKey()).color(Colors.VARIABLE_VALUE)
                success(" gesetzt.")
            }
        }
    }
}